vmm/src/vstate.rs

```diff
@@ -261,19 +278,124 @@ impl Vcpu {
+            Ok(run) => match run {
+                VcpuExit::IoIn(addr, data) => {
+                    self.io_bus.read(u64::from(addr), data);
+                    METRICS.vcpu.exit_io_in.inc();
+                    Ok(())
+                }
+                VcpuExit::IoOut(addr, data) => {
+                    if addr == MAGIC_IOPORT_SIGNAL_GUEST_BOOT_COMPLETE
+                        && data[0] == MAGIC_VALUE_SIGNAL_GUEST_BOOT_COMPLETE
+                    {
+                        super::Vmm::log_boot_time(self.create_ts.clone());
+                    }
+                    self.io_bus.write(u64::from(addr), data);
+                    METRICS.vcpu.exit_io_out.inc();
+                    Ok(())
+                }
+                VcpuExit::MmioRead(addr, data) => {
+                    self.mmio_bus.read(addr, data);
+                    METRICS.vcpu.exit_mmio_read.inc();
+                    Ok(())
+                }
+                VcpuExit::MmioWrite(addr, data) => {
+                    self.mmio_bus.write(addr, data);
+                    METRICS.vcpu.exit_mmio_write.inc();
+                    Ok(())
+                }
+                VcpuExit::Hlt => {
+                    info!("Received KVM_EXIT_HLT signal");
+                    Err(Error::VcpuUnhandledKvmExit)
+                }
+                VcpuExit::Shutdown => {
+                    info!("Received KVM_EXIT_SHUTDOWN signal");
+                    Err(Error::VcpuUnhandledKvmExit)
+                }
+                // Documentation specifies that below kvm exits are considered
+                // errors.
+                VcpuExit::FailEntry => {
+                    METRICS.vcpu.failures.inc();
+                    error!("Received KVM_EXIT_FAIL_ENTRY signal");
+                    Err(Error::VcpuUnhandledKvmExit)
+                }
+                VcpuExit::InternalError => {
+                    METRICS.vcpu.failures.inc();
+                    error!("Received KVM_EXIT_INTERNAL_ERROR signal");
+                    Err(Error::VcpuUnhandledKvmExit)
+                }
+                r => {
+                    METRICS.vcpu.failures.inc();
+                    // TODO: Are we sure we want to finish running a vcpu upon
+                    // receiving a vm exit that is not necessarily an error?
+                    error!("Unexpected exit reason on vcpu run: {:?}", r);
+                    Err(Error::VcpuUnhandledKvmExit)
+                }
+            },
+            // The unwrap on raw_os_error can only fail if we have a logic
+            // error in our code in which case it is better to panic.
+            Err(ref e) => {
+                match e.raw_os_error().unwrap() {
+                    // Why do we check for these if we only return EINVAL?
+                    libc::EAGAIN | libc::EINTR => Ok(()),
+                    _ => {
+                        METRICS.vcpu.failures.inc();
+                        error!("Failure during vcpu run: {:?}", e);
```

Comment:
`e` is an `std::io::Error`, please use the `{}` formatter instead.

Reply:
done